core/src/main/scala/kafka/log/Log.scala

```diff
@@ -255,19 +261,21 @@ case object SnapshotGenerated extends LogStartOffsetIncrementReason {
 @threadsafe
 class Log(@volatile private var _dir: File,
           @volatile var config: LogConfig,
+          val segments: LogSegments,
```

Comment:
Can we pass `LogComponents` as a single field instead of initializing with the respective fields? This makes the code easy to read, constructor simpler with less vars.

Reply:
I see your point. It's good thinking. But the main reason I didn't take this route was that some of the fields of `LogComponents` are immutable within the context of the `LogLoader`, but are mutable by `Log`. Examples: `logStartOffset`, `recoveryPoint` and `leaderEpochCache ` are defined as `var` only in `Log`. It seemed hard to explain if we made it a `var` inside `LogComponents` just because the `Log` class wants to mutate them in the future.

Reply:
What I meant was we can have `Log` takes the argument as immutable `LogComponents` and it can initialize the vars inside `Log` with the respective fields from `LogComponents`. This will also set the right access of these vars by not giving write access by default. 

```
class Log(@volatile private var _dir: File,
          @volatile var config: LogConfig,
          val segments: LogSegments,
          val logComponents: LogComponents,
          scheduler: Scheduler,
          brokerTopicStats: BrokerTopicStats,
          val time: Time,
          val maxProducerIdExpirationMs: Int,
          val producerIdExpirationCheckIntervalMs: Int,
          val topicPartition: TopicPartition,
          logDirFailureChannel: LogDirFailureChannel,
          @volatile var topicId: Option[Uuid],
          val keepPartitionMetadataFile: Boolean = true) extends Logging with KafkaMetricsGroup {



  @volatile private var logStartOffset: Long = logComponents.logStartOffset
  @volatile private var recoveryPoint: Long = logComponents.recoveryPoint
  @volatile private var nextOffsetMetadata: LogOffsetMetadata = logComponents.nextOffsetMetadata
  @volatile var leaderEpochCache: Option[LeaderEpochFileCache] = logComponents.leaderEpochCache
  private val producerStateManager: ProducerStateManager = logComponents.producerStateManager
```

Reply:
Sounds good. I'll give this a shot.

Reply:
I took a stab at this. Things changed a bit after I did the refactoring that Jun suggested [here](https://github.com/apache/kafka/pull/10478#discussion_r610933966). In the latest version of the PR, `LogComponents` has been renamed to a class called `LoadedLogOffsets` which now only contains (logStartOffset, recoveryPoint, nextOffsetMetadata). It seemed reasonable to just pass these in as 3 separate parameters into `Log` since the `LeaderEpochFileCache` and `ProducerStateManager` are no longer inside the `LoadedLogOffsets` object, so those have to be passed separately anyway.

Reply:
Sure, that looks reasonable to me.